"reflect: call of reflect.Value.IsNil on struct Value" for struct error type

The following code panics, as `NilError` is attempting to check `IsNil` on a non-pointer struct:

```
type structError struct{}

func (structError) Error() string {
	return "this is structError"
}

func TestStructError(t *testing.T) {
	var err error = structError{}
	assert.NilError(t, err)
}
```

```
--- FAIL: TestStructError (0.00s)
panic: reflect: call of reflect.Value.IsNil on struct Value [recovered]
	panic: reflect: call of reflect.Value.IsNil on struct Value

goroutine 21 [running]:
testing.tRunner.func1(0xc000110300)
	/usr/lib/go/src/testing/testing.go:874 +0x3a3
panic(0x59a640, 0xc000096120)
	/usr/lib/go/src/runtime/panic.go:679 +0x1b2
reflect.Value.IsNil(...)
	/usr/lib/go/src/reflect/value.go:1073
gotest.tools/v3/assert.assert(0x60d4c0, 0xc000110300, 0xc00007ef00, 0x5dec38, 0x5a0360, 0x748a08, 0x0, 0x0, 0x0, 0x0)
	/home/jake/go/pkg/mod/gotest.tools/v3@v3.0.1/assert/assert.go:123 +0x773
gotest.tools/v3/assert.NilError(0x60d4c0, 0xc000110300, 0x609d80, 0x748a08, 0x0, 0x0, 0x0)
	/home/jake/go/pkg/mod/gotest.tools/v3@v3.0.1/assert/assert.go:249 +0x101
some.tld/baderr.TestStructError(0xc000110300)
	/home/jake/testproj/baderr/err_test.go:36 +0x60
testing.tRunner(0xc000110300, 0x5df150)
	/usr/lib/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/lib/go/src/testing/testing.go:960 +0x350
FAIL	some.tld/baderr	0.009s
FAIL
```

The struct doesn't need to be empty; it can have values. Related to #167/#171.

However, I think #171 needs to be reverted anyway (#185), because a typed nil error is not "nil".

--- a/assert/assert.go
+++ b/assert/assert.go
@@ -119,12 +119,8 @@ func assert(
 		return true
 
 	case error:
-		// Handle nil structs which implement error as a nil error
-		if reflect.ValueOf(check).IsNil() {
-			return true
-		}
-		msg := "error is not nil: "
-		t.Log(format.WithCustomMessage(failureMessage+msg+check.Error(), msgAndArgs...))
+		msg := failureMsgFromError(check)
+		t.Log(format.WithCustomMessage(failureMessage+msg, msgAndArgs...))
 
 	case cmp.Comparison:
 		success = runComparison(t, argSelector, check, msgAndArgs...)
@@ -177,6 +173,15 @@ func logFailureFromBool(t TestingT, msgAndArgs ...interface{}) {
 	}
 
 	t.Log(format.WithCustomMessage(failureMessage+msg, msgAndArgs...))
+}
+
+func failureMsgFromError(err error) string {
+	// Handle errors with non-nil types
+	v := reflect.ValueOf(err)
+	if v.Kind() == reflect.Ptr && v.IsNil() {
+		return fmt.Sprintf("error is not nil: error has type %T", err)
+	}
+	return "error is not nil: " + err.Error()
 }
 
 func boolFailureMessage(expr ast.Expr) (string, error) {

--- a/assert/assert_test.go
+++ b/assert/assert_test.go
@@ -115,32 +115,70 @@ func TestAssertWithComparisonAndExtraMessage(t *testing.T) {
 	expectFailNowed(t, fakeT, "assertion failed: oops, not good: extra stuff true")
 }
 
-type customError struct{}
+type customError struct {
+	field bool
+}
 
 func (e *customError) Error() string {
+	// access a field of the receiver to simulate the behaviour of most
+	// implementations, and test handling of non-nil typed errors.
+	e.field = true
 	return "custom error"
 }
 
-func TestNilErrorSuccess(t *testing.T) {
-	fakeT := &fakeTestingT{}
+func TestNilError(t *testing.T) {
+	t.Run("nil interface", func(t *testing.T) {
+		fakeT := &fakeTestingT{}
+		var err error
+		NilError(fakeT, err)
+		expectSuccess(t, fakeT)
+	})
 
-	var err error
-	NilError(fakeT, err)
-	expectSuccess(t, fakeT)
+	t.Run("nil literal", func(t *testing.T) {
+		fakeT := &fakeTestingT{}
+		NilError(fakeT, nil)
+		expectSuccess(t, fakeT)
+	})
 
-	NilError(fakeT, nil)
-	expectSuccess(t, fakeT)
+	t.Run("interface with non-nil type", func(t *testing.T) {
+		fakeT := &fakeTestingT{}
+		var customErr *customError
+		NilError(fakeT, customErr)
+		expected := "assertion failed: error is not nil: error has type *assert.customError"
+		expectFailNowed(t, fakeT, expected)
+	})
 
-	var customErr *customError
-	NilError(fakeT, customErr)
-	expectSuccess(t, fakeT)
+	t.Run("non-nil error", func(t *testing.T) {
+		fakeT := &fakeTestingT{}
+		NilError(fakeT, fmt.Errorf("this is the error"))
+		expectFailNowed(t, fakeT, "assertion failed: error is not nil: this is the error")
+	})
+
+	t.Run("non-nil error with struct type", func(t *testing.T) {
+		fakeT := &fakeTestingT{}
+		err := structError{}
+		NilError(fakeT, err)
+		expectFailNowed(t, fakeT, "assertion failed: error is not nil: this is a struct")
+	})
+
+	t.Run("non-nil error with map type", func(t *testing.T) {
+		fakeT := &fakeTestingT{}
+		var err mapError
+		NilError(fakeT, err)
+		expectFailNowed(t, fakeT, "assertion failed: error is not nil: ")
+	})
 }
 
-func TestNilErrorFailure(t *testing.T) {
-	fakeT := &fakeTestingT{}
+type structError struct{}
 
-	NilError(fakeT, fmt.Errorf("this is the error"))
-	expectFailNowed(t, fakeT, "assertion failed: error is not nil: this is the error")
+func (structError) Error() string {
+	return "this is a struct"
+}
+
+type mapError map[int]string
+
+func (m mapError) Error() string {
+	return m[0]
 }
 
 func TestCheckFailure(t *testing.T) {

